### PR TITLE
Substitution compose

### PR DESCRIPTION
**Pull request by Julow**
The first commit is moving code to avoid cross dependencies.
The implementation is very naive and to be tested correctly.

~~This PR is based on top of https://github.com/jonludlam/odoc/pull/2 and the test is still passing.~~

TODO:
- [x] Write unit tests

### CODE REVIEW

**Comment by Julow** [comment]
I think this is read for reviews.

File: src/xref2/component.ml [hunk 1]
```diff
@@ -31,26 +31,6 @@ module IdentMap = Map.Make (struct
...
-    | Some x -> x
-    | None ->
-        let v = x.get () in
-        x.v <- Some v;
-        v
-
-  let put : (unit -> 'a) -> 'a t =
-   fun f ->
-    if !eager then { v = Some (f ()); get = f } else { v = None; get = f }
-end
```
Comment (jonludlam) [inline src/xref2/component.ml]: I'm still not keen on moving this module to Subst - we'll be delaying strengthening soon too, for instance.

File: src/xref2/test/subst/test_compose.ml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    All_subst.(test_module [ [ ("A", "B") ]; [ ("C", "D") ]; [ ("A", "C") ] ]);
+    All_subst.(test_module [ [ ("A", "B"); ("C", "D") ]; [ ("D", "E"); ("F", "C") ]; [ ("B", "F") ] ]);
+    All_subst.(test_module [ [ ("A", "B") ]; [ ("A", "C") ] ]);
+
+    Subst_type_expr.(test
+        [
+          [ type_ "a", Tuple [ constr "b"; constr "c" ] ];
+          [ type_ "c", constr "d" ];
+        ]
+        ~test_cases:[ constr "a" ]);
```
Comment (jonludlam) [inline src/xref2/test/subst/test_compose.ml]: It'd be quite nice to see the expectation here too - I'm guessing we're expecting `a` to be `b*d` right?

File: src/xref2/tools.ml [hunk 3]
```diff
@@ -1302,10 +1303,10 @@ and fragmap_type :
             match item with
             | Component.Signature.Module (id, r, m)
               when Ident.Name.module_ id = ModuleName.to_string name ->
-                let m = Component.Delayed.get m in
+                let m = Subst.Delayed.get_module m in (* TODO: Do we need to apply subst ? *)
```
Comment (jonludlam) [inline src/xref2/tools.ml]: Yes, we do :-)

File: src/xref2/component.ml (hunk 1 above)
Comment (Julow) [reply src/xref2/component.ml]: This module is not moved, it's removed.
`Subst.Delayed` doesn't do delayed computation, it's the corresponding module for `'a Substitution.delayed`.

File: src/xref2/tools.ml (hunk 3 above)
Comment (Julow) [reply src/xref2/tools.ml]: Thanks :)

File: src/xref2/test/subst/test_compose.ml (hunk 2 above)
Comment (Julow) [reply src/xref2/test/subst/test_compose.ml]: That's a good idea, I'll add that.
A mistake I made while rebasing prevented the tests from running. This test is failing now, I'm investigating.

File: src/xref2/component.ml (hunk 1 above)
Comment (jonludlam) [reply src/xref2/component.ml]: Ah, I misread this as `Subst.delayed` - even so though, it's still within the `Substitution` module, which is going to be odd when we add in strengthening.

Comment (Julow) [reply src/xref2/component.ml]: By delaying strengthening, do you mean lazy computation or composed substitutions ?
In the first case, `Subst.Delayed` doesn't do that. I reused the name for something unrelated.
In the second, it is necessary to depend on `Subst`.

**Comment by Julow** [comment]
I've fixed the `type_replacement` bug.

File: src/xref2/component.ml (hunk 1 above)
Comment (jonludlam) [reply src/xref2/component.ml:53]: I'm imagining the delayed stuff as a list of transforms that we're applying to a more-or-less immutable base. Substs are one such transformation, strengthening is another.

Comment (Julow) [reply src/xref2/component.ml:53]: If there is code/types to share between subst and strengthening, I think we should do it later.
For now, `Component.Delayed` is removed and `Subst.Delayed` don't replace it. Should I bring `Component.Delayed` back ?

Comment (jonludlam) [reply src/xref2/component.ml:53]: It's not so much that there is code to share as that conceptually we're moving in the right direction. In any case, it's something we can sort out later as you say. LGTM :-)

**Pull request merged**